numpy/core/src/umath/ufunc_object.c

```diff
@@ -5428,7 +5430,12 @@ ufunc_outer(PyUFuncObject *ufunc, PyObject *args, PyObject *kwds)
     if (tmp == NULL) {
         return NULL;
     }
-    ap1 = (PyArrayObject *) PyArray_FromObject(tmp, NPY_NOTYPE, 0, 0);
+    if (PyObject_IsInstance(tmp, numpy_matrix)) {
```

Comment:
I also spent some time looking at ways to probe for attributes unique to `np.matrix`, in the hopes of keeping this check confined to the C level as much as possible.

Those with more matrix experience may have a better idea.

Reply:
Do we care enough about `matrix` *sub*classes to do the `IsInstance` rather than a direct type check? (Assuming the latter is fast also for pure python classes.)

Reply:
This seems wrong. Why are we special-casing `matrix`, which we want to deprecate anyway, deep inside the ufunc machinery? Matrix is no more special than any other ndarray subclass, and should be using a protocol to avoid getting here in the first place.

Reply:
Because if matrix isn't handled specially somehow, it will create a behavior change here. Trying to implement a correct `__array_ufunc__ ` for a deprecated type is probably not worth the time vs just hacking around it until it can be removed.